File: ios/App/Components/ProjectComponent.swift
import Foundation
import AppGraphqlApi

/// Project mutations. Reads flow through ProjectsPageComponent's feed; this
/// component owns writes and cross-cutting side effects (refresh + alerts).
@MainActor
final class ProjectComponent {
  /// Returns true when the project was created (so sheets can dismiss).
  func createProject(name: String, description: String?) async -> Bool {
    let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmedName.isEmpty else {
      reportError("Enter a project name.")
      return false
    }
    let trimmedDescription = description?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    do {
      _ = try await gql.createProject(
        input: CreateProjectInput(
          idempotencyKey: UUID().uuidString,
          fields: CreateProjectFields(
            name: trimmedName,
            description: trimmedDescription.isEmpty ? .none : .some(trimmedDescription)
          )
        )
      )
      await components.projectsPage.refresh(withLoading: false)
      reportSuccess("Project created.")
      return true
    } catch {
      reportError(error.localizedDescription)
      return false
    }
  }

  func archiveProject(projectId: String) async -> Bool {
    do {
      _ = try await gql.updateProject(
        input: UpdateProjectInput(
          objectId: projectId,
          idempotencyKey: UUID().uuidString,
          fields: UpdateProjectFields(doArchive: .some(true))
        )
      )
      await components.projectsPage.refresh(withLoading: false)
      reportSuccess("Project archived.")
      return true
    } catch {
      reportError(error.localizedDescription)
      return false
    }
  }

  private func reportError(_ message: String) {
    let trimmed = message.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else { return }
    store.sessionStore.reportError(trimmed)
    store.appAlertStore.activeAlert = AppAlertStore.AlertMessage(
      id: "project-error-\(trimmed)",
      message: trimmed,
      isError: true
    )
  }

  private func reportSuccess(_ message: String) {
    store.sessionStore.reportSuccess(message)
    store.appAlertStore.activeAlert = AppAlertStore.AlertMessage(
      id: "project-success-\(message)",
      message: message,
      isError: false
    )
  }
}

File: android/app/src/main/kotlin/com/baseapp/android/components/ProjectComponent.kt
package com.baseapp.android.components

import com.apollographql.apollo.api.Optional
import com.baseapp.android.graphql.generated.type.CreateProjectFields
import com.baseapp.android.graphql.generated.type.CreateProjectInput
import com.baseapp.android.graphql.generated.type.UpdateProjectFields
import com.baseapp.android.graphql.generated.type.UpdateProjectInput
import com.baseapp.android.store.AppAlertStore
import java.util.UUID

/**
 * Project mutations. Reads flow through ProjectsPageComponent's feed; this
 * component owns writes and cross-cutting side effects (refresh + alerts).
 */
class ProjectComponent {
    /** Returns true when the project was created (so sheets can dismiss). */
    suspend fun createProject(name: String, description: String?): Boolean {
        val trimmedName = name.trim()
        if (trimmedName.isEmpty()) {
            reportError("Enter a project name.")
            return false
        }
        val trimmedDescription = description?.trim() ?: ""
        return try {
            gql.createProject(
                CreateProjectInput(
                    idempotencyKey = UUID.randomUUID().toString(),
                    fields = CreateProjectFields(
                        name = trimmedName,
                        description = if (trimmedDescription.isEmpty()) {
                            Optional.absent()
                        } else {
                            Optional.present(trimmedDescription)
                        },
                    ),
                )
            )
            components.projectsPage.refresh(withLoading = false)
            reportSuccess("Project created.")
            true
        } catch (error: Exception) {
            reportError(error.message ?: "Could not create the project.")
            false
        }
    }

    suspend fun archiveProject(projectId: String): Boolean {
        return try {
            gql.updateProject(
                UpdateProjectInput(
                    objectId = projectId,
                    idempotencyKey = UUID.randomUUID().toString(),
                    fields = UpdateProjectFields(doArchive = Optional.present(true)),
                )
            )
            components.projectsPage.refresh(withLoading = false)
            reportSuccess("Project archived.")
            true
        } catch (error: Exception) {
            reportError(error.message ?: "Could not archive the project.")
            false
        }
    }

    private fun reportError(message: String) {
        val trimmed = message.trim()
        if (trimmed.isEmpty()) {
            return
        }
        store.sessionStore.reportError(trimmed)
        store.appAlertStore.setActiveAlert(
            AppAlertStore.AlertMessage(id = "project-error-$trimmed", message = trimmed, isError = true)
        )
    }

    private fun reportSuccess(message: String) {
        store.sessionStore.reportSuccess(message)
        store.appAlertStore.setActiveAlert(
            AppAlertStore.AlertMessage(id = "project-success-$message", message = message, isError = false)
        )
    }
}
